File: leetcode/0052.go
package leetcode

func totalNQueens(n int) int {
	ans := 0
	var dfs func(row, col, l, r int)
	dfs = func(row, col, l, r int) {
		if row == n {
			ans++
			return
		}
		for cur := ((1 << n) - 1) &^ (col | l | r); cur > 0; cur &= (cur - 1) {
			pos := cur & -cur
			dfs(row+1, col|pos, (l|pos)>>1, (r|pos)<<1)
		}
	}
	dfs(0, 0, 0, 0)
	return ans
}

File: leetcode/0052.cpp
class Solution {
public:
    int totalNQueens(int n) {
        auto ans = 0;
        auto dfs = [&](auto&& dfs, int row, int col, int l, int r) {
            if (row == n) {
                ans++;
                return;
            }
            for (auto cur = ((1 << n) - 1) & (~(col | l | r)); cur;
                 cur &= (cur - 1)) {
                int pos = cur & (-cur);
                dfs(dfs, row + 1, col | pos, (l | pos) << 1, (r | pos) >> 1);
            }
        };
        dfs(dfs, 0, 0, 0, 0);
        return ans;
    }
};